contracts/multisig/src/signing.rs

```diff
@@ -79,15 +77,32 @@ fn signers_weight(
         .sum()
 }
 
+fn update_session_state(
+    store: &mut dyn Storage,
+    session: &mut SigningSession,
+    key: &Key,
+    block_height: u64,
+) -> StdResult<()> {
+    // weight must be loaded outside `if`, otherwise it could cause gas cost inconsistency among signers
+    let weight = signers_weight(store, session.id.u64(), &key.snapshot)?;
+    if session.state == MultisigState::Pending && weight >= key.snapshot.quorum.into() {
+        session.state = MultisigState::Completed {
+            completed_at: block_height,
+        };
+    }
+
+    SIGNING_SESSIONS.save(store, session.id.u64(), session)
+}
```

Comment:
I'd like to keep storage management separate from decision logic. Consider loading and saving separately

Reply:
Agree with Chris

Reply:
did some restructuring, please take a look again